modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/database/H2TreeInlineObjectDetector.java

```diff
@@ -68,26 +85,61 @@
                 continue;
             }
 
-            if (r.getValue(ih.columnIndex()) == ValueNull.INSTANCE)
+            Value val = r.getValue(ih.columnIndex());
+
+            if (val == ValueNull.INSTANCE)
                 return false;
 
             int type = PageUtils.getByte(pageAddr, off + fieldOff);
 
-            if (type == 0) {
-                inlineObjSupported = false;
+            //We can have garbage in memory and need to compare data.
+            if (type == Value.JAVA_OBJECT) {
+                int len = PageUtils.getShort(pageAddr, off + fieldOff + 1);
 
-                return true;
-            }
-            else if (type == Value.JAVA_OBJECT) {
-                inlineObjSupported = true;
+                boolean isFull = (len & 0x8000) == 0;
+
+                len = len & 0x7FFF;
+
+                byte[] originalObjBytes = val.getBytesNoCopy();
+
+                // read size more then available space or more then origin length
+                if(len > inlineSize - fieldOff - 3 || len > originalObjBytes.length){
+
+                    inlineObjectSupportedDecision(false, "lenght is big " + len);
+
+                    return true;
+                }
+
+                // compare full object
+                if(isFull) {
+                    int compRes = ih.compare(pageAddr, off + fieldOff, inlineSize - fieldOff, val, (o1, o2) -> 0);
+
+                    inlineObjectSupportedDecision(compRes == 0, "full compare");
+
+                    return true;
+                }
+
+
+                // try compare byte by byte for partial inlined object.
+                byte[] inlineBytes = PageUtils.getBytes(pageAddr, off + fieldOff + 3, len);
+
+                for (int i = 0; i < len; i++) {
+                    if(inlineBytes[i] == originalObjBytes[i])
+                        continue;
+
+                    inlineObjectSupportedDecision(false,  i + " byte compare");
+
+                    return true;
+                }
+
+                inlineObjectSupportedDecision(true, len + " bytes compared");
 
                 return true;
             }
-            else {
-                assert type == Value.UNKNOWN;
 
-                return false;
-            }
+            inlineObjectSupportedDecision(false, "inline type " + type);
+
+            return true;
```

Comment:
It does not neccessarily mean that we finished here. Type byte can be `Value.UNKNOWN` for that entry, but next one can be successfully inlined.

Reply:
UNKNOWN can be put only for case when value can't fit to inlineSize. It mean last field which could be inlined

Reply:
I meant not field/column but entry (row). Consider an index (String, JAVA_OBJECT) it's inlined entries:
1: long string, UNKNOWN for java object.
2: short string, inlined java object

`return` statement in a commented line decides that inline for java objects is not supported while it should.

Reply:
fixed